### PR TITLE
MetainfoFetcher and file lists

### PR DESCRIPTION
**Pull request by ElegantMonkey**
A few changes from #440:
- Renamed `FilesizeFetcher` to `MetainfoFetcher`.
- Updated logic so that it gets the metainfo of both torrents missing file sizes or missing file lists.
- Show file lists on the View Torrent page.
- Read file list from uploaded .torrent files.
- Make failed fetches be able to be fetched again, with a exponential cooldown.

~~I couldn't test this because my connection is really trash right now~~ Tested it, seems to work fine albeit a bit slow, might need to experiment with Timeout/Queue size. PR'ing for thoughts/suggestions.

### CODE REVIEW

**Comment by ewhal** [comment]
Seems to be working but I haven't tested it on any batches yet

File: service/torrent/metainfoFetcher/metainfoFetcher.go [hunk 1]
```diff
@@ -150,12 +152,28 @@ func (fetcher *MetainfoFetcher) gotResult(r Result) {
...
+	if fetcher.failCooldown < 0 {
+		return
+	}
+
+	now := time.Now()
+	for id, failTime := range fetcher.failedOperations {
+		if failTime.Add(time.Duration(fetcher.failCooldown) * time.Second).Before(now) {
+			log.Infof("Torrent TID %d gone through cooldown, removing from failures")
+			// Deleting keys inside a loop seems to be safe.
+			delete(fetcher.failedOperations, id)
```
Comment (majestrate) [inline service/torrent/metainfoFetcher/metainfoFetcher.go]: pretty sure this will crash

Comment (ElegantMonkey) [inline service/torrent/metainfoFetcher/metainfoFetcher.go]: I was afraid too, but after some Googling it seems to be safe. [https://github.com/golang/go/issues/9926](url)

Comment (majestrate) [inline service/torrent/metainfoFetcher/metainfoFetcher.go]: is the map only accessed from 1 goroutine ?

Comment (ElegantMonkey) [inline service/torrent/metainfoFetcher/metainfoFetcher.go]: This map is only accessed by functions called from `MetadataFetcher.run()`, which is a single goroutine. The spawned goroutines don't access it, AFAIK. Could use a Mutex just for safety, though.

**Pull request merged**